### PR TITLE
Bandwdith fixes

### PR DESCRIPTION
**Pull request by AndriusA**
Resolves https://github.com/brave/brave-browser/issues/8256
Resolves https://github.com/brave/brave-browser/issues/8568
Resolves https://github.com/brave/brave-browser/issues/8359

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [x] Linux
  - [x] macOS
  - [x] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [x] Linux
  - [x] macOS
  - [x] Windows
- [x] Verified that all lint errors/warnings are resolved (`npm run lint`)
- [x] Ran `git rebase master` (if needed).
- [ ] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protection-Mode
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: components/brave_perf_predictor/browser/perf_predictor_tab_helper_browsertest.cc [hunk 1]
```diff
@@ -138,6 +140,7 @@ IN_PROC_BROWSER_TEST_F(PerfPredictorTabHelperTest, NewNavigationStoresSavings) {
   GURL second_url =
       embedded_test_server()->GetURL("example.com", "/blocking.html");
   ui_test_utils::NavigateToURL(browser(), second_url);
+  ASSERT_TRUE(ExecuteScript(contents, "addImage('logo.png')"));
```
Comment (iefremov) [inline components/brave_perf_predictor/browser/perf_predictor_tab_helper_browsertest.cc]: mind replacing all `ExecuteScript` with `EvalJs/ExecJs` (which is a recommended API now)?

File: components/brave_perf_predictor/browser/p3a_bandwidth_savings_permanent_state.cc [hunk 2]
```diff
@@ -52,18 +60,13 @@ void P3ABandwidthSavingsPermanentState::AddSavings(uint64_t delta) {
 
 base::Optional<uint64_t>
```
Comment (iefremov) [inline components/brave_perf_predictor/browser/p3a_bandwidth_savings_permanent_state.cc]: now there is no need to return `Optional`

File: components/brave_perf_predictor/browser/p3a_bandwidth_savings_permanent_state.h [hunk 3]
```diff
@@ -7,7 +7,9 @@
 #define BRAVE_COMPONENTS_BRAVE_PERF_PREDICTOR_BROWSER_P3A_BANDWIDTH_SAVINGS_PERMANENT_STATE_H_
 
 #include <list>
+#include <memory>
 
+#include "base/time/clock.h"
```
Comment (iefremov) [inline components/brave_perf_predictor/browser/p3a_bandwidth_savings_permanent_state.h]: it seems like forward declaration is enough for `base::Clock`

File: components/brave_perf_predictor/browser/p3a_bandwidth_savings_permanent_state_unittest.cc [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_perf_predictor/browser/p3a_bandwidth_savings_permanent_state.h"
+
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "base/compiler_specific.h"
```
Comment (iefremov) [inline components/brave_perf_predictor/browser/p3a_bandwidth_savings_permanent_state_unittest.cc]: hmm, what is it for? And the next one

File: components/brave_perf_predictor/browser/p3a_bandwidth_savings_tracker.h [hunk 5]
```diff
@@ -7,6 +7,9 @@
 #define BRAVE_COMPONENTS_BRAVE_PERF_PREDICTOR_BROWSER_P3A_BANDWIDTH_SAVINGS_TRACKER_H_
 
 #include <cstdint>
+#include <memory>
+
+#include "base/time/clock.h"
```
Comment (iefremov) [inline components/brave_perf_predictor/browser/p3a_bandwidth_savings_tracker.h]: replace with forward declaration?

File: components/brave_perf_predictor/browser/p3a_bandwidth_savings_tracker.h [hunk 6]
```diff
@@ -16,6 +19,8 @@ namespace brave_perf_predictor {
 class P3ABandwidthSavingsTracker {
  public:
   explicit P3ABandwidthSavingsTracker(PrefService* user_prefs);
+  P3ABandwidthSavingsTracker(PrefService* user_prefs,
```
Comment (iefremov) [inline components/brave_perf_predictor/browser/p3a_bandwidth_savings_tracker.h]: pls add a comment saying this one is for testing

File: components/brave_perf_predictor/browser/p3a_bandwidth_savings_tracker_unittest.cc [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/brave_perf_predictor/browser/p3a_bandwidth_savings_tracker.h"
+
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "base/compiler_specific.h"
```
Comment (iefremov) [inline components/brave_perf_predictor/browser/p3a_bandwidth_savings_tracker_unittest.cc]: pls check for unneeded headers

File: components/brave_perf_predictor/browser/p3a_bandwidth_savings_tracker.cc [hunk 8]
```diff
@@ -62,11 +70,11 @@ void P3ABandwidthSavingsTracker::RegisterPrefs(PrefRegistrySimple* registry) {
 
 void P3ABandwidthSavingsTracker::StoreSavingsHistogram(uint64_t savings_bytes) {
   int bucket = 0;
-  uint64_t total_mb = static_cast<uint64_t>(savings_bytes / 1024 / 1024);
-  int counter = 0;
+  // divide by 1024*1024 = 2^20 to convert bytes -> MB
+  uint64_t total_mb = savings_bytes >> 20;
   for (const auto& bucketUpperBound : kBandwidthSavingsBuckets) {
     if (total_mb > bucketUpperBound)
-      bucket = counter + 1;
```
Comment (iefremov) [inline components/brave_perf_predictor/browser/p3a_bandwidth_savings_tracker.cc]: omg! :(

File: components/brave_perf_predictor/browser/p3a_bandwidth_savings_permanent_state_unittest.cc [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+ public:
+  P3ABandwidthSavingsPermanentStateTest() : clock_(new base::SimpleTestClock) {
+    P3ABandwidthSavingsTracker::RegisterPrefs(pref_service_.registry());
+    state_ = std::make_unique<P3ABandwidthSavingsPermanentState>(
+        &pref_service_, std::unique_ptr<base::Clock>(clock_));
+    clock_->SetNow(base::Time::Now());
+  }
+
+ protected:
+  base::SimpleTestClock* clock_;
```
Comment (iefremov) [inline components/brave_perf_predictor/browser/p3a_bandwidth_savings_permanent_state_unittest.cc]: this is memleak, mb just store it by value? Or replace with unique_ptr

File: components/brave_perf_predictor/browser/p3a_bandwidth_savings_permanent_state.h [hunk 10]
```diff
@@ -25,6 +27,8 @@ namespace brave_perf_predictor {
 class P3ABandwidthSavingsPermanentState {
  public:
   explicit P3ABandwidthSavingsPermanentState(PrefService* user_prefs);
+  P3ABandwidthSavingsPermanentState(PrefService* user_prefs,
```
Comment (iefremov) [inline components/brave_perf_predictor/browser/p3a_bandwidth_savings_permanent_state.h]: pls mark this as "for testing only"

**Comment by iefremov** [comment]
lgtm with nits

**Comment by AndriusA** [comment]
CI errors seem to be in lint script itself:

```
21:15:07  Traceback (most recent call last):
21:15:07    File "/home/ubuntu/jenkins/workspace/browser-build-pr_bandwdith-fixes/scripts/lint.py", line 84, in <module>
21:15:07      sys.exit(main(sys.argv[1:]))
21:15:07    File "/home/ubuntu/jenkins/workspace/browser-build-pr_bandwdith-fixes/scripts/lint.py", line 50, in main
21:15:07      change = cl.GetChange(git_common.get_or_create_merge_base(cl.GetBranch(), options.base_branch))
21:15:07  TypeError: GetChange() takes exactly 3 arguments (2 given)
```

and `npm audit deps`, having rebased yesterday just before pushing for CI, not sure how those should be resolved?

**Comment by bsclifton** [comment]
Added missing 1.7 milestone from this PR

**Pull request merged**